runtime/vm/exceptionsupport.c

```diff
@@ -1291,3 +1291,22 @@ setIllegalAccessErrorFinalFieldSet(J9VMThread *currentThread, UDATA isStatic, J9
 	setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALACCESSERROR, msg);
 	j9mem_free_memory(msg);
 }
+
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+void
+setCRIUSingleThreadModeJVMCheckpointException(J9VMThread *vmThread, U_32 moduleName, U_32 messageNumber)
+{
+	PORT_ACCESS_FROM_VMC(vmThread);
+	const char *msg = NULL;
+
+	/* If no custom NLS message was specified, use the generic one */
+	if ((0 == moduleName) && (0 == messageNumber)) {
+		moduleName = J9NLS_VM_CRIU_SINGLETHREADMODE_JVMCHECKPOINTEXCEPTION__MODULE;
+		messageNumber = J9NLS_VM_CRIU_SINGLETHREADMODE_JVMCHECKPOINTEXCEPTION__ID;
+	}
+	msg = OMRPORT_FROM_J9PORT(PORTLIB)->nls_lookup_message(OMRPORT_FROM_J9PORT(PORTLIB), J9NLS_DO_NOT_PRINT_MESSAGE_TAG | J9NLS_DO_NOT_APPEND_NEWLINE, moduleName, messageNumber, NULL);
+
+	/* set org.eclipse.openj9.criu.JVMCheckpointException */
+	setCurrentExceptionUTF(vmThread, J9VMCONSTANTPOOL_ORGECLIPSEOPENJ9CRIUJVMCHECKPOINTEXCEPTION, msg);
```

Comment:
you need to detect if you are single thread mode on the checkpoint side, or the restore side. That should determine which exception should be thrown.

We'll need to add a new flag in criusupport.cpp that indicates which side we are on

Reply:
Can `vm->portLibrary->checkpointRestoreTimeDelta` serve the purpose, `0` means checkpoint side, otherwise restore instead?

Reply:
what about multiple checkpoints?

Reply:
`checkpointRestoreTimeDelta` needs to be reset to `0` before entering single thread mode, this value itself is not expected to support multiple checkpoints.

Reply:
`vm->portLibrary->checkpointRestoreTimeDelta = 0;` will have to be moved to before the java hooks are run

Reply:
Moved `vm->portLibrary->checkpointRestoreTimeDelta = 0` before `jvmCheckpointHooks()`.
Set `JVMCheckpointException` or `RestoreException` according to if `checkpointRestoreTimeDelta` is `0`.